test/test-udp-reuseport.c

```diff
@@ -0,0 +1,258 @@
+  buf->len = (int) suggested_size;
+}
+
+static void ticktack(uv_timer_t* timer) {
+  int done = 0;
+
+  ASSERT(timer == &thread_timer_handle1 || timer == &thread_timer_handle2);
+
+  uv_mutex_lock(&mutex);
+  if (received == MAX_UDP_DATAGRAMS) {
+    done = 1;
+  }
+  uv_mutex_unlock(&mutex);
+
+  if (done) {
+    uv_close((uv_handle_t*) timer, NULL);
+    if (timer->loop == thread_loop1)
+      uv_close((uv_handle_t*) &thread_handle1, NULL);
+    if (timer->loop == thread_loop2)
+      uv_close((uv_handle_t*) &thread_handle2, NULL);
+  }
+}
+
+static void on_recv(uv_udp_t* handle,
+                    ssize_t nr,
+                    const uv_buf_t* buf,
+                    const struct sockaddr* addr,
+                    unsigned flags) {
+  ASSERT_OK(flags);
+  ASSERT(handle == &thread_handle1 || handle == &thread_handle2);
+
+  ASSERT_GE(nr, 0);
+
+  if (nr == 0) {
+    ASSERT_NULL(addr);
+    free(buf->base);
+    return;
+  }
+
+  ASSERT_NOT_NULL(addr);
+  ASSERT_EQ(5, nr);
+  ASSERT(!memcmp("Hello", buf->base, nr));
+  free(buf->base);
+
+  if (handle->loop == thread_loop1)
+    thread_loop1_recv++;
+
+  if (handle->loop == thread_loop2)
+    thread_loop2_recv++;
+
+  uv_mutex_lock(&mutex);
+  received++;
+  uv_mutex_unlock(&mutex);
+}
+
+static void on_send(uv_udp_send_t* req, int status) {
+  ASSERT_OK(status);
+  ASSERT_PTR_EQ(req->handle->loop, main_loop);
+
+  if (++sent == MAX_UDP_DATAGRAMS)
+    uv_close((uv_handle_t*) req->handle, NULL);
+}
+
+static void run_event_loop(void* arg) {
+  int r;
+  uv_loop_t* loop = (uv_loop_t*) arg;
+  ASSERT(loop == thread_loop1 || loop == thread_loop2);
+
+  /* Notify the main thread to start sending data. */
+  uv_sem_post(&semaphore);
+  r = uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT_OK(r);
+}
+
+static void bind_socket_and_prepare_recv(uv_loop_t* loop, uv_udp_t* handle) {
+  struct sockaddr_in addr;
+  int r;
+
+  ASSERT_OK(uv_ip4_addr("127.0.0.1", TEST_PORT, &addr));
+
+  r = uv_udp_init(loop, handle);
+  ASSERT_OK(r);
+
+  r = uv_udp_bind(handle, (const struct sockaddr*) &addr, UV_UDP_REUSEPORT);
+  ASSERT_OK(r);
+
+  r = uv_udp_recv_start(handle, alloc_cb, on_recv);
+  ASSERT_OK(r);
+}
+
+TEST_IMPL(udp_reuseport) {
+  struct sockaddr_in addr;
+  uv_buf_t buf;
+  int r;
+  int i;
+
+  r = uv_mutex_init(&mutex);
+
+  r = uv_sem_init(&semaphore, 0);
+  ASSERT_OK(r);
+
+  main_loop = uv_default_loop();
+  ASSERT_NOT_NULL(main_loop);
+
+  /* Create receiving socket per event loop. */
+  thread_loop1 = uv_loop_new();
+  ASSERT_NOT_NULL(thread_loop1);
+  bind_socket_and_prepare_recv(thread_loop1, &thread_handle1);
+  uv_timer_init(thread_loop1, &thread_timer_handle1);
+  uv_timer_start(&thread_timer_handle1, ticktack, 0, 10);
+
+  thread_loop2 = uv_loop_new();
+  ASSERT_NOT_NULL(thread_loop2);
+  bind_socket_and_prepare_recv(thread_loop2, &thread_handle2);
+  uv_timer_init(thread_loop2, &thread_timer_handle2);
+  uv_timer_start(&thread_timer_handle2, ticktack, 0, 10);
```

Comment:
Personally I'd call the `bind_socket_and_prepare_recv()` methods from the actual thread the loop is running on.

Reply:
I guess we can do that. But if we do that, it seems that we may also want to move `uv_timer_init` and `uv_timer_start` there as well. Is that an actual benefit or just a preference?

Reply:
Yes those too. It's a personal preference in the sense that most of libuv methods aren't thread-safe so just to be sure they're always called from the loop thread it's a good idea in order to avoid problems. Of course in this case it's not an issue as the threads haven't been started yet. Feel free to dismiss the comment though.

Reply:
Yes, both ways should have the same effect. But indeed your suggestion will make the code more feasible for each thread to initialize their own data. I'd like to go with that for `tcp_reuseport` and `udp_reuseport`.